resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigurationStepsOrchestrator.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.k8s.submit
+
+import org.apache.spark.SparkConf
+import org.apache.spark.deploy.k8s.Config._
+import org.apache.spark.deploy.k8s.ConfigurationUtils
+import org.apache.spark.deploy.k8s.Constants._
+import org.apache.spark.deploy.k8s.submit.steps._
+import org.apache.spark.launcher.SparkLauncher
+import org.apache.spark.util.SystemClock
+
+/**
+ * Constructs the complete list of driver configuration steps to run to deploy the Spark driver.
+ */
+private[spark] class DriverConfigurationStepsOrchestrator(
+    namespace: String,
+    kubernetesAppId: String,
+    launchTime: Long,
+    mainAppResource: Option[MainAppResource],
+    appName: String,
+    mainClass: String,
+    appArgs: Array[String],
+    submissionSparkConf: SparkConf) {
+
+  // The resource name prefix is derived from the application name, making it easy to connect the
+  // names of the Kubernetes resources from e.g. kubectl or the Kubernetes dashboard to the
+  // application the user submitted. However, we can't use the application name in the label, as
+  // label values are considerably restrictive, e.g. must be no longer than 63 characters in
+  // length. So we generate a separate identifier for the app ID itself, and bookkeeping that
+  // requires finding "all pods for this application" should use the kubernetesAppId.
+  private val kubernetesResourceNamePrefix =
```

Comment:
Yes, it needs to be unique. Changed to use `UUID.nameUUIDFromBytes` based on `launchTime`. @mccheah any concern about this?

Reply:
UUIDs are a bit long which leads to name length problems since Kubernetes requires some names to be < 63 characters. I would agree that a UUID would be better if we can get away with it.

Reply:
I cant see rest of the context of this conversation - but UUID will be 36 odd bytes, well within the 63 char limit ?

Reply:
I think the biggest constraint here is labels, which cannot be longer than 63 characters. API object names have a limit of 253 characters. Are we using the prefix in any label?